### ISSUE TITLE
Conversion from a protected_function_result to an object

### ISSUE THREAD

**Issue by Philpax** [ThePhD/sol2#236]
As far as I can see, there's no direct way of converting a `protected_function_result` to an `object`; there are no assignment operators, nor any cast operators. 

However, constructing an `object` using the state + stack index of the `protected_function_result` seems to work fine; I added a constructor to `object` to facilitate this in my local version, but am not sure if there are any potential ramifications (double pops, perhaps?)

My rationale for conversion is being able to use `object`'s greater feature set (i.e. `is`). An alternative might be to extend the interface of `protected_function_result` to match that of `object`'s.

**Comment by ThePhD** [comment]
![It's already done](http://i.imgur.com/iJiF9sI.png)

There's an implicit cast operator for _all types of T_ in `proxy_base`, from which `protected_function_result` derives. This is mentioned in the docs, and if you tried it the conversion would've Just Worked™.

In general, you should try the code you think would work in Sol. If it doesn't work, then file a bug report and I'll basically take the necessary hour or so making it possible, if it's not possible already.

**Comment by Philpax** [comment]
Ah, whoops - it's just one specific case I was having issues with: assigning a `protected_function_result` to an already-existing `object`. You're right about being able to construct an `object` from a `protected_function_result` - sorry about not checking that properly! (Was trying `auto obj = sol::object(fnResult);` and `sol::object obj(fnResult);`, neither of which trigger the cast because they're both constructors)

For assignment, the conversion is ambiguous:

```cpp
#include "sol.hpp"
#include <iostream>

int main() {
	sol::state lua;
	lua.open_libraries();

	lua.script("function f() return 24 end");

	sol::protected_function pf = lua["f"];
	sol::object o = pf();
	std::cout << o.as<int>() << std::endl;
	// Ambiguity error
	o = pf();

	return 0;
}
```

Errors:
```
1>  main.cpp
1>main.cpp(14): error C2593: 'operator =' is ambiguous
1>  sol.hpp(9313): note: could be 'sol::basic_object<sol::reference> &sol::basic_object<sol::reference>::operator =(base_t &&)'
1>          with
1>          [
1>              base_t=sol::reference
1>          ]
1>  sol.hpp(9312): note: or       'sol::basic_object<sol::reference> &sol::basic_object<sol::reference>::operator =(const base_t &)'
1>          with
1>          [
1>              base_t=sol::reference
1>          ]
1>  sol.hpp(9311): note: or       'sol::basic_object<sol::reference> &sol::basic_object<sol::reference>::operator =(sol::basic_object<sol::reference> &&)'
1>  sol.hpp(9310): note: or       'sol::basic_object<sol::reference> &sol::basic_object<sol::reference>::operator =(const sol::basic_object<sol::reference> &)'
1>  main.cpp(14): note: while trying to match the argument list '(sol::object, sol::protected_function_result)'
```

However, following your advice to check out [the docs](http://sol2.readthedocs.io/en/latest/api/proxy.html), I used the templated `get` method for something that works:
```cpp
int main() {
	sol::state lua;
	lua.open_libraries();

	lua.script("function f() return 24 end");

	sol::protected_function pf = lua["f"];
	sol::object o = pf();
	std::cout << o.as<int>() << std::endl;
	o = pf().get<sol::object>();
	std::cout << o.as<int>() << std::endl;

	return 0;
}
```

So this is not an impediment, but somewhat counter-intuitive. No biggie if you decide against doing anything further with it, though. It's probably worth making it clearer in the docs that `operator T` and `get<T>` (the latter more so) will work fine with types derived from `sol::reference`, especially `sol::object`; I was under the impression they were solely for retrieving C++ objects.

Thanks in advance, and sorry for wasting your time earlier!

**Comment by ThePhD** [comment]
It's fine: you're not wasting my time by posting this, it just helps me to know what other kinds of examples and stuff I should be making.

Also, to void having to do `o = pf().get<sol::object>();`, I've added a few new constructors and assignment operators. The kind of code you posted is what helps me figure out what users need and where I've derped (I'm only one person working ont his, you know). Descriptions sometimes work, but ultimately end up in me guessing and doesn't really help me open up my perspective about what _code_ I'm failing to accomodate.

Again, don't be sorry, and thanks for posting the more detailed code!

Added here: https://github.com/ThePhD/sol2/releases/tag/v2.14.9
